compiler/cpp/src/generate/t_java_generator.cc

```diff
@@ -3967,14 +3972,42 @@ string t_java_generator::function_signature_async(t_function* tfunction, bool us
  * @param tfunction Function definition
  * @return String of rendered function definition
  */
-string t_java_generator::function_signature_future(t_function* tfunction, string prefix) {
+string t_java_generator::function_signature_future(t_function* tfunction, int params_futures_mask, string prefix) {
   t_type* ttype = tfunction->get_returntype();
   std::string return_type = type_name(ttype, true);
   std::string fn_name = get_rpc_method_name(tfunction->get_name());
-  std::string arg_list = argument_list(tfunction->get_arglist());
+  std::string arg_list = futures_argument_list(tfunction->get_arglist(), params_futures_mask);
   return "CompletableFuture<" + return_type + "> " + prefix + fn_name + "(" + arg_list + ")";
 }
 
+/**
+ * Renders a comma separated field list, with type names. The types are
+ * modified to use a future of the argument type according to the mask
+ * given.
+ */
+string t_java_generator::futures_argument_list(t_struct* tstruct, int params_futures_mask) {
+  string result = "";
+
+  const vector<t_field*>& fields = tstruct->get_members();
+  vector<t_field*>::const_iterator f_iter;
+  int i = 0;
+  for (f_iter = fields.begin(); f_iter != fields.end(); ++f_iter, ++i) {
+    if (i > 0) {
+      result += ", ";
+    }
+    string type;
+    if (params_futures_mask & (1 << i)) {
+      // The mask indicates to use a future for this argument
+      type = "CompletableFuture<" + type_name((*f_iter)->get_type(), true) + ">";
```

Comment:
This is a consumer, so let's consume the most generic thing we can; Future<T> instead of CompletableFuture<T>

There is one thing we should probably discuss here though; what do we do about numeric type promotion? For example, if the function expects Future<Long>, and we feed it a Future<Integer>, should that compile? Right now, it won't; to do that we'd either need even more overrides, or we'd have to accept a Future<Number>. Future<Number> has a different problem; you can now feed a long to a function that expects a short, and you'll get a runtime rather than compile-time error. I'm happy postponing this problem for now.